### PR TITLE
Add coupled IAU RT test

### LINKED ISSUE #1613: Update MOM_input templates to be usable for global applications GFS and GEFS
## Description
The `global-workflow` carries around its set of `MOM_input_template_$OCNRES`
It has been a long desired feature to use the templates from the components themselves.  In this case, the ufs-weather-model

## Solution
- update `MOM_input_template_$OCNRES` to allow templating of variables that are currently hard-wired for RT.
- update config files for impacted RTs to provide values used in RT that are currently hard-wired.

## Alternatives
The `global-workflow` carry around its own set of `MOM_input_template_$OCNRES`

## Related to
This work is a prerequisite for https://github.com/NOAA-EMC/global-workflow/issues/1312

@JessicaMeixner-NOAA will determine the timeline and assignment for this work.

### LINKED ISSUE #1856: create coupled IAU RT test case
## Description
The GFSv17 will run with IAU in the coupled configuration. A test case need to set up with IAU turned on for both atmosphere and ocean to ensure that IAU and all the component outputs are working properly.

## Solution
1) Create C384 ICs for the coupled case with IAU increments for atmosphere and ocean (dummy ocean increments, see https://github.com/NOAA-EMC/global-workflow/issues/1776).
2) Create test case with IAU turned on for fv3atm and mom6. 

## Related to
- global workflow issue https://github.com/NOAA-EMC/global-workflow/issues/1776

### PR DESCRIPTION
**Pull request by junwang-noaa**
## PR Author Checklist:

- [x] I have linked PR's from all sub-components involved in section below. 
- [ ] I am confirming reviews are completed in ALL sub-component PR's.
- [ ] I have run the full RT suite on either Hera/Cheyenne AND have attached the log to this PR below this line:
  - LOG: 
- [ ] I have added the list of all failed regression tests to "Anticipated changes" section.
- [ ] I have filled out all sections of the template.

## Description
A coupled IAU RT test cpld_control_gfsv17_iau is created for GFSV17 testing.

### Associated UFSWM Issue to close

- Closes #1856
- Closes #1613 

### Subcomponent Pull Requests

None

### Blocking Dependencies

None

### Subcomponents involved:
- [ ] AQM
- [ ] CDEPS
- [ ] CICE
- [ ] CMEPS
- [ ] CMakeModules
- [ ] FV3
- [ ] GOCART
- [ ] HYCOM
- [ ] MOM6
- [ ] NOAHMP
- [ ] WW3
- [ ] stochastic_physics
- [x] none

### Input data
- [ ] No changes are expected to input data.
- [x] Changes are expected to input data:
  - [x] New input data.
  - [ ] Updated input data.

### Regression Tests:
- [x] No changes are expected to any regression test:  baseline is needed for the new test, no results change for current tests
- [ ] Changes are expected to the following tests:

<details><summary>Tests effected by changes in this PR:</summary>

</details>

### Libraries

- [x] Not Needed
- [ ] Needed
  - [ ] Create separate issue in [JCSDA/spack-stack](https://github.com/JCSDA/spack-stack) asking for update to library. Include library name, library version.
  - [ ] Add issue link from JCSDA/spack-stack following this item 

<details><summary>Code Managers Log</summary>

- [ ] This PR is up-to-date with the top of all sub-component repositories except for those sub-components which are the subject of this PR.
- [ ] Move new/updated input data on RDHPCS Hera and propagate input data changes to all supported systems.
  - [ ] N/A

### Testing Log:
- RDHPCS
  - [x] Hera
  - [ ] Orion
  - [ ] Hercules
  - [ ] Jet
  - [ ] Gaea
  - [ ] Cheyenne
- WCOSS2
  - [ ] Dogwood/Cactus
  - [x] Acorn
- CI
  - [ ] Completed
- opnReqTest
  - [ ] N/A
  - [ ] Log attached to comment
</details>

### CODE REVIEW

File: tests/fv3_conf/cpld_control_run.IN [hunk 1]
```diff
@@ -96,6 +96,10 @@ else
        RFILE_OLD=$(basename $RFILE)
           mv -f $RFILE INPUT/"${RFILE_OLD//${RESTART_FILE_PREFIX}./}"
      done
+     echo "IAU_OFFSET=$IAU_OFFSET, FHROT=$FHROT"
+     if [[ $IAU_OFFSET == 6 ]] && [[ $FHROT -le 6 ]]; then
+       cp /scratch1/NCEPDEV/nems/Jun.Wang/nems/vlab/inc/c96/atm/fv_increment*.nc ./INPUT/.
```
Comment (junwang-noaa) [inline tests/fv3_conf/cpld_control_run.IN]: @jkbk2004 I need UFS code managers' help to copy these data /scratch1/NCEPDEV/nems/Jun.Wang/nems/vlab/inc/c96/atm/fv_increment*.nc to input-data-20221101/FV3_input_data/INPUT_L127_mx100/. Thanks

File: tests/fv3_conf/cpld_control_run.IN [hunk 2]
```diff
@@ -105,6 +109,12 @@ else
       cp ../${DEP_RUN}${SUFFIX}/RESTART/${RESTART_FILE_PREFIX}.MOM.res_3.nc ./INPUT/MOM.res_3.nc
      else
       cp ../${DEP_RUN}${SUFFIX}/RESTART/${RESTART_FILE_PREFIX}.MOM.res.nc ./INPUT/MOM.res.nc
+#      if [[ ${IAU_OFFSET} == 6 ]] && [[ $FHROT <= 6 ]] ; then
+       if [[ ${IAU_OFFSET} == 6 ]] ; then
+       if [[ $FHROT -le 6 ]] ; then
+        cp /scratch1/NCEPDEV/nems/Jun.Wang/nems/vlab/inc/c96/ocn/mom6_increment.nc ./INPUT/mom6_increment.nc
```
Comment (junwang-noaa) [inline tests/fv3_conf/cpld_control_run.IN]: @jkbk2004 I need code managers' help to copy this data /scratch1/NCEPDEV/nems/Jun.Wang/nems/vlab/inc/c96/ocn/mom6_increment.nc  to: input-data-20221101/MOM6_IC/.  on all the platforms. Thanks

**Comment by junwang-noaa** [comment]
The full RT passed except the missing baseline for the new test. The RT log is at:

/scratch1/NCEPDEV/nems/Jun.Wang/nems/vlab/20230817/ufs-weather-model/tests/logs

**Comment by junwang-noaa** [comment]
@jiandewang would you please check if the mom6 output from the IAU test at:

/scratch1/NCEPDEV/stmp2/Jun.Wang/FV3_RT/rt_248142/cpld_control_gfsv17_iau_intel ? Thanks

**Comment by jiandewang** [comment]
> @jiandewang would you please check if the mom6 output from the IAU test at:
> 
> /scratch1/NCEPDEV/stmp2/Jun.Wang/FV3_RT/rt_248142/cpld_control_gfsv17_iau_intel ? Thanks

@junwang-noaa I had a quick look on ocean output, nothing strange

**Comment by JessicaMeixner-NOAA** [comment]
@jiandewang @NeilBarton-NOAA - This PR is testing the IAU capability which includes wanting the averaging buckets to be zerod out at FHR=0 when we have FHROT=3.  Can you confirm if the ocean/ice models are correctly doing this zero-ing out or if this remain open issues?

**Comment by jiandewang** [comment]
> /scratch1/NCEPDEV/stmp2/Jun.Wang/FV3_RT/rt_248142/cpld_control_gfsv17_iau_intel



> @jiandewang @NeilBarton-NOAA - This PR is testing the IAU capability which includes wanting the averaging buckets to be zerod out at FHR=0 when we have FHROT=3. Can you confirm if the ocean/ice models are correctly doing this zero-ing out or if this remain open issues?

@JessicaMeixner-NOAA I am not quite understanding what you want me to check. All I see is there are two ocean outout files:
ocn_2021_03_22_12.nc and ocn_2021_03_22_18.nc. Model started at 6z and saved every 6hr output

**Comment by jiandewang** [comment]
> > /scratch1/NCEPDEV/stmp2/Jun.Wang/FV3_RT/rt_248142/cpld_control_gfsv17_iau_intel
> 
> > @jiandewang @NeilBarton-NOAA - This PR is testing the IAU capability which includes wanting the averaging buckets to be zerod out at FHR=0 when we have FHROT=3. Can you confirm if the ocean/ice models are correctly doing this zero-ing out or if this remain open issues?
> 
> @JessicaMeixner-NOAA I am not quite understanding what you want me to check. All I see is there are two ocean outout files: ocn_2021_03_22_12.nc and ocn_2021_03_22_18.nc. Model started at 6z and saved every 6hr output

what is "wanting the averaging buckets to be zerod out at FHR=0 when we have FHROT=3." ?

**Comment by jiandewang** [comment]
> > > /scratch1/NCEPDEV/stmp2/Jun.Wang/FV3_RT/rt_248142/cpld_control_gfsv17_iau_intel
> > 
> > 
> > > @jiandewang @NeilBarton-NOAA - This PR is testing the IAU capability which includes wanting the averaging buckets to be zerod out at FHR=0 when we have FHROT=3. Can you confirm if the ocean/ice models are correctly doing this zero-ing out or if this remain open issues?
> > 
> > 
> > @JessicaMeixner-NOAA I am not quite understanding what you want me to check. All I see is there are two ocean outout files: ocn_2021_03_22_12.nc and ocn_2021_03_22_18.nc. Model started at 6z and saved every 6hr output
> 
> what is "wanting the averaging buckets to be zerod out at FHR=0 when we have FHROT=3." ?

I guess this test is on FV3 side it is doing this kind of IAU but not on ocean side, @junwang-noaa can you confirm on this ?

**Comment by DeniseWorthen** [comment]
@junwang-noaa For CICE, my initial read is that this is doing what you want for the CICE history files. In your run, in ``ice_diag.d``, I see that the first history file ``iceh_06h.2021-03-22-43200.nc`` is written after 15 timesteps and 15*720secs = 3 hours. The next file ``iceh_06h.2021-03-22-64800.nc`` is written after 30 timesteps (6 hours). 

I will do a test and see if the fields in the history files appear to be 3 and 6 hour averages as expected.

**Comment by JessicaMeixner-NOAA** [comment]
That's great news @DeniseWorthen     

@jiandewang we're basically expecting to see what @DeniseWorthen mentioned. FHROT means that we change when we start the forecast relative to the forecast hour 0.  We can hop on a quick call tomorrow tand I can try to explain more if needed.

**Comment by jiandewang** [comment]
> That's great news @DeniseWorthen
> 
> @jiandewang we're basically expecting to see what @DeniseWorthen mentioned. FHROT means that we change when we start the forecast relative to the forecast hour 0. We can hop on a quick call tomorrow tand I can try to explain more if needed.

now I know what's your purpose. But for ocean all I see is 12z and 18z output. I think diag_table doesn't have this capability but ice has a way to control that.

**Comment by junwang-noaa** [comment]
@DeniseWorthen Thanks for confirmation. I think ice is OK. We do not expect users to use any averaged fields in the first ice output iceh_06h.2021-03-22-43200.nc, which is corresponding to atm f00 at 12z cycle. The next history file iceh_06h.2021-03-22-64800.nc has correct 6 hour average.

**Comment by junwang-noaa** [comment]
@jiandewang if I understand correct, the ocn files should also doing the same thing as ice does. But please check the SST file:

 /scratch1/NCEPDEV/stmp2/Jun.Wang/FV3_RT/rt_248142/cpld_control_gfsv17_iau_intel/SST_2021_03_23.nc

which is a daily mean. I assume we will not output it in marine DA (@guillaumevernieres) step, but will be output in GFSv17 forecast and this daily mean starts from the initial forecast time (start time), which is 6 hourly before the current cycle. E.g. in this case, the SST mean is from 2023062206- 2023062318, while this is a 12Z cycle run and we need it to be 2023062212-2023022306.

**Comment by junwang-noaa** [comment]
@JessicaMeixner-NOAA would you please confirm the wave fields coming out correctly? Thanks

**Comment by JessicaMeixner-NOAA** [comment]
The wave output is as requested which is instantaneous point and gridded files every 3 hours.

**Comment by DeniseWorthen** [comment]
@jiandewang @junwang-noaa  The ``SST_2021_03_23.nc`` file contains all missing values for the SST. I think this is because FMS does not have a complete "24 hour" set of values to accumulate and average. I'm not sure if this file is actually needed.

I can't remember how the 3-d files are named. Is the timestamp the end of the averaging period, so that ``ocn_2021_03_22_18.nc`` contain the 6-hour mean of hours 13,14,15,16,17,18?

**Comment by jiandewang** [comment]
@DeniseWorthen you are right, this setting only run 18hr so SST file has all missing value. For 6hr 3D the file name timestamp is using end of time.

**Comment by DeniseWorthen** [comment]
@jiandewang Thanks for the clarification. In that case, I am not confident that we're getting the expected values in the ocean file but I can't yet be entirely sure. What I did was write mediator history file 1-hour averages, averaged them over the intervals and compared that to what the model has for the same interval. For the ice model, I saw differences O\~10-8 but I'm seeing differences more like O~10-1 w/ the MOM6 file for SST.

**Comment by junwang-noaa** [comment]
@DeniseWorthen @jiandewang Thanks. I think the ocn file is in the mid-time. Please see ocn output from 24 hr forecast with a cold start:
/scratch1/NCEPDEV/stmp2/Jun.Wang/FV3_RT/rt_182886/cpld_control_gfsv17_intel

**Comment by jiandewang** [comment]
> @DeniseWorthen @jiandewang Thanks. I think the ocn file is in the mid-time. Please see ocn output from 24 hr forecast with a cold start: /scratch1/NCEPDEV/stmp2/Jun.Wang/FV3_RT/rt_182886/cpld_control_gfsv17_intel

sorry my bad. ocn file is the middle time. In g-w Prototype and HR runs we rename them to ending time in the script. In rt.sh we don't do this renaming step. I mixed rt.sh and g-w.

**Comment by DeniseWorthen** [comment]
@jiandewang Ok, my brain isn't working today! What hours are used to create the ocn output that is labeled "18"?

**Comment by jiandewang** [comment]
> @jiandewang Ok, my brain isn't working today! What hours are used to create the ocn output that is labeled "18"?

19z-24z

**Comment by DeniseWorthen** [comment]
That can't be right? Wouldn't that be the hour=21 file?

**Comment by junwang-noaa** [comment]
I think it's 16-21

**Comment by jiandewang** [comment]
> I think it's 16-21

sorry my bad again, my brain isn't working today

**Comment by DeniseWorthen** [comment]
In the CICE history files, everything seems consistent. The 43200 file has time bounds = (0.125,0.25) in units of ``days since 2021-03-22 06:00:00``. So that is a 3 hour average (hours 9:12). The next file has bounds (0.25,0.50) so that is a 6 hour average (hours 12:18). 

In MOM6, the _12 file has time bounds (3,9) in units of ``hours since 2021-03-22 06:00:00`` so that would be clock hours (9:15). The _18 file has time bounds of (9,15) so clock hours of (15:21). 

Is this the right average for MOM6? It is not the same averaging window as CICE but I don't know which is right.

**Comment by junwang-noaa** [comment]
Hera RT log file is at 
/scratch1/NCEPDEV/nems/Jun.Wang/nems/vlab/20230817/ufs-weather-model/tests/logs/RegressionTests_hera.log

**Comment by junwang-noaa** [comment]
@jkbk2004 The PR is ready.

**Comment by zach1221** [comment]
@junwang-noaa We're almost ready to begin testing against this PR. Can you please sync up your branch with develop?

**Comment by junwang-noaa** [comment]
@zach1221 Done

**Comment by zach1221** [comment]
This should be labelled as baseline change due to the new test being created.

**Comment by BrianCurtis-NOAA** [comment]
Is this new input data on WCOSS2/Acorn already?

Also, we can use the current baseline directory, adding a new test should not need a new Bldir.

**Comment by junwang-noaa** [comment]
No, not on wcoss2 yet.

**Comment by zach1221** [comment]
> Is this new input data on WCOSS2/Acorn already?
> 
> Also, we can use the current baseline directory, adding a new test should not need a new Bldir.

Ok, understood. I'll flip the label back.

**Comment by zach1221** [comment]
My first test run failed on Hera for the new test because of missing baselines. Do we need to change the directories in the cpld_control_run.IN file so it points to the new locations for the new input data in input-data-20221101? Currently still points to Jun's scratch directory.

**Comment by DeniseWorthen** [comment]
@zach1221 Yes, assuming the file has been added to the inputdata on all platforms, @junwang-noaa needs to make that update in ``cpld_control_run.IN``. Or, can you push to her branch?

**Comment by zach1221** [comment]
> @zach1221 Yes, assuming the file has been added to the inputdata on all platforms, @junwang-noaa needs to make that update in `cpld_control_run.IN`. Or, can you push to her branch?

Yes, the new input data has been distributed across the platforms. Except for wcoss2 I guess.

**Comment by jkbk2004** [comment]
> > @zach1221 Yes, assuming the file has been added to the inputdata on all platforms, @junwang-noaa needs to make that update in `cpld_control_run.IN`. Or, can you push to her branch?
> 
> Yes, the new input data has been distributed across the platforms. Except for wcoss2 I guess.

Increment files were rsynced ok acorss all RDHPCS.

**Comment by DeniseWorthen** [comment]
@zach1221 Jun asked me to update the directory in the PR. Can you please try to generate a baseline for the new test now? Hopefully I got it right, but I did not test first.

**Comment by zach1221** [comment]
> @zach1221 Jun asked me to update the directory in the PR. Can you please try to generate a baseline for the new test now? Hopefully I got it right, but I did not test first.

Ok, thanks. I wasn't certain what it should be, so was testing a change on my end. I'll try yours and let you know how it goes.

**Review by aerorahul — DISMISSED** [review]
Changes look good to me.
Thank you for synchronizing the `MOM_input_templates` from the global-workflow.
Please link the issue #1613 with this PR.  That issue will be resolved with this merge.

**Comment by zach1221** [comment]
@junwang-noaa we're now finished with testing. If you're able to close the two unresolved conversations, then we can begin the merging process.

**Comment by junwang-noaa** [comment]
Done

**Comment by DeniseWorthen** [comment]
@junwang-noaa I meant to ask---I know at some point you were testing an updated FMS in relation to this PR? Is an updated FMS required to have the IAU working properly for MOM6 output?

**Comment by junwang-noaa** [comment]
That's right. I am waiting for an FMS release so that we can get correct MOM6 output in IAU configuration. It may take some time, I am still checking the timeline of the FMS release

**Review by DeniseWorthen — APPROVED** [review]

**Review by sadeghitabas — APPROVED** [review]

**Pull request merged**